rfcs/syntax-safe-navigation-operator.md

```diff
@@ -0,0 +1,137 @@
+
+The long example would turn into:
+
+```lua
+logDogName(getLogger(), dog?.name, dog?.owner?:getDisplayName())
+```
+
+Failing the nil-safety check early would make the entire expression nil, for instance `dog?.body.legs` would resolve to `nil` if `dog` is nil, rather than resolve `dog?.body` into nil, then turning into `nil.legs`.
+
+The list of valid operators to follow the safe navigation operator would be:
+
+```lua
+dog?.name --[[ is the same as ]] if dog == nil then nil else dog.name
+dog?.getName() --[[ is the same as ]] if dog == nil then nil else dog.getName()
+dog?:getName(args) --[[ is the same as ]] if dog == nil then nil else dog:getName(args)
+```
+
+When using safe navigation to call a function, the short circuiting will prevent the arguments from being evaluated in the case of nil.
+
+```lua
+-- Will NOT call getValue() if `object` is nil
+object?.doSomething(getValue())
+```
+
+The short-circuiting is limited within the expression.
+
+```lua
+dog?.owner.name -- This will return nil if `dog` is nil
+(dog?.owner).name -- `(dog?.owner)` resolves to nil, of which `name` is then indexed. This will error at runtime if `dog` is nil.
+
+dog?.legs + 3 -- `dog?.legs` is resolved on its own, meaning this will error at runtime if it is nil (`nil + 3`)
+```
+
+The operator must be used in the context of either a call or an index, and so:
+
+```lua
+local value = x?
+```
+
+...would be invalid syntax. 
+
+This syntax would be based on expressions, and not identifiers, meaning that `(x or y)?.call()` would be valid syntax.
+
+### Type
+If the expression is typed as an optional, then the resulting type would be the final expression, also optional. Otherwise, it'll just be the resulting type if `?` wasn't used.
+
+```lua
+local optionalObject: { name: string }?
+local optionalObjectName = optionalObject?.name -- resolves to `string?`
+
+local nonOptionalObject: { name: string }
+local nonOptionalObjectName = nonOptionalObject?.name -- resolves to `string`
+```
+
+### Assignment
+`x?.y = z` is not supported, and will be reported as a syntax error.
+
+## Drawbacks
+
+As with all syntax additions, this adds complexity to the parsing of expressions, and the execution of cancelling the rest of the expression could prove challenging.
+
+Furthermore, with the proposed syntax, it might lock off other uses of `?` within code (and not types) for the future as being ambiguous.
+
+## Alternatives
+
+Doing nothing is an option, as current standard if-checks already work, as well as the `and` trick in other use cases, but as shown before this can create some hard to read code, and nil values are common enough that the safe navigation operator is welcome.
+
+Supporting optional calls/indexes, such as `x?[1]` and `x?()`, while not out of scope, are likely too fringe to support, while adding on a significant amount of parsing difficulty, especially in the case of shorthand function calls, such as `x?{}` and `x?""`.

```

Comment:
I wouldn't say `x?[index]` is a fringe case. It seems especially useful for checking if something exists at an index and doing something with the value.

```lua
local function sendMessage(object, message, ...)
  object?.__definition?[message](object, ...)
end
```

Reply:
This line was added by request from Luau maintainers, I think it's a reasonable request but perhaps not something that needs to be in this initial RFC.